### ISSUE TITLE
Questions: what does it mean to get 507 code from envoy?

### ISSUE THREAD

**Issue by Keramblock** [envoyproxy/envoy#15211]
*Title*: *Questions: what does it mean to get 507 code from envoy?*

*Description*:
Hi, I just get response code: 507 with response text: "exceeded request buffer limit while retrying upstream". And I just cant understand what happend, because the only trace of that error I can find is link to the place where error was raised in source code:
https://github.com/envoyproxy/envoy/blob/main/source/common/router/router.cc#L687 
But I still just cant understand what exactly happenned there.
Could you please explain it?

[optional *Relevant Links*:]
https://github.com/envoyproxy/envoy/blob/main/source/common/router/router.cc#L687

**Comment by mattklein123** [comment]
cc @alyssawilk

**Comment by alyssawilk** [comment]
Envoy has bounds on how much it'll buffer for retries, which are enforced a bit more strictly than how much Envoy will buffer for original requests.
If Envoy is waiting on a retry (which often have back-off to not overwhelm the upstream) and more data arrives than the configured limit (see https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/flow_control on flow control limits) it will give up buffering and return an error.

**Comment by lambdai** [comment]
@alyssawilk  A random thought: is it possible to switch to something like `StopIterationAndWatermark` when waiting on retry?

**Comment by alyssawilk** [comment]
We might be able to do better with flow control limits, but fundamentally Envoy has always had limits on how much it'll buffer for retries.

Imagine you're processing a really large POST.  If the upstream disconnects early on during the POST, Envoy will retry.  If it then gets more and more data, it will eventually stop buffering POST data, and if the upstream disconnects late in the POST, Envoy returns an error rather than retrying.

Now add in the retry delay.  Imagine upstream fails early on in the POST and Envoy has buffered *almost* as much as it's willing to buffer for a retry attempt.  I'm not sure if we'll kick in flow control (we should! and it's worth looking into to minimize this corner case!).  But even if we kick in flow control, back-off is not immediate for HTTP/2, so we'll continue to get a bit more data. If that kicks us over the retry buffer threshold, it acts as if we'd had an upstream failure after more of the post and just fails to retry.  We could chose to treat this as we do with most flow control and allow a little more than the limit to be buffered, but it's a bunch of high risk work and I'm not convinced anyone's going to tackle it to handle the race condition where we always have to handle the case that retries won't happen due to large bodies.  If that all makes sense?

**Comment by lambdai** [comment]
1 round trip to propagate the per h2 stream window size (plus the OS socket buffer sitting between) could cause the retry buffer go beyond the buffer limit. 

Thank you for the explanation of this complex problem!

**Comment by nitgoy** [comment]
Hi @alyssawilk, @lambdai. We're seeing this issue as well and after understanding the issue were also wondering if we could kick in flow control and stop reading from downstream during retry delay for such a case. That way envoy stops buffering the data during the backoff delay when it's not actively streaming that data to any upstream. I understand that it may not work perfectly and envoy might still return an error in many cases. But sounds like it should provide _some_ improvement to the error rate, especially when the backoff delay is relatively large.

**Comment by alyssawilk** [comment]
does https://github.com/envoyproxy/envoy/pull/32359 help?

**Comment by fl0Lec** [comment]
Hi, nope https://github.com/envoyproxy/envoy/pull/32359 does not help for this issue.
I guess we could pause the read on the downstream connection with the `onDecoderFilterAboveWriteBufferHighWatermark` and temporary increase the buffer limit when such a case present itself.

There was just a concern from @KBaichoo that data might remain in the proxy: https://github.com/envoyproxy/envoy/pull/32359#discussion_r1526455610. 

I am not so familiar yet with about the impact of flow control but would be happy to get more context about it and explore it.

**Comment by nitgoy** [comment]
> does #32359 help?
No, it doesn't. As I understand, #32359 is talking about avoiding resetting the retries when the buffer gets full _before_ receiving the failure from upstream. The problem discussed in the current issue is of 507s when the buffer gets full _after_ receiving failure from upstream but before issuing the request to the next upstream host. I guess both would be nice fixes and might potentially share parts of code changes relating to downstream backpressure. But for us 507 is more of a concern since it directly affects our SLA whereas #32359 only causes suboptimal retry behavior.

**Comment by alyssawilk** [comment]
gotcha.  yeah there's probably other suboptimal side cases - if you can put together a patch I'd be happy to review but unfortunately I'm too overloaded to pick it up myself any time soon :-(
